### PR TITLE
[WIP] Refactor `SearchBar`, `SourceSearch` and `Autocomplete` and clean up function search UI

### PR DESCRIPTION
**Pull request by wldcordeiro**
Associated Issue: #2151 

### Summary of Changes

* [x] consolidate the input fields in `SearchBar` and `Autocomplete` into `SearchInput`
* [x] Break the result list from `Autocomplete` into its own component `ResultList` to be used in `SearchBar`
* [ ] Clean up the UI/UX for function searching

Test plan:

- [ ] Open debugger
- [ ] Test the project file search <kbd>ctrl</kbd>/<kbd>cmd</kbd>+P works as before
- [ ] Test out that normal editor searches <kbd>ctrl</kbd>/<kbd>cmd</kbd>+F work as before
- [ ] Test out that searching for functions by toggling the button in the bottom bar works as expected
- [ ] Test out using the keyboard shortcut for function search <kbd>ctrl</kbd>/<kbd>cmd</kbd>+<kbd>shift</kbd>+O works

### Screenshots/Videos

screenshot_20170226_041508

### CODE REVIEW

File: src/utils/result-list.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  const itemHeight = resultsEl.children[0].clientHeight;
+  const numVisible = resultsHeight / itemHeight;
+  const positionsToScroll = index - numVisible + 1;
+  const itemOffset = resultsHeight % itemHeight;
+  const scroll = positionsToScroll * (itemHeight + 2) + itemOffset;
+
+  resultsEl.scrollTop = Math.max(0, scroll);
+}
+
+function handleKeyDown(e: SyntheticKeyboardEvent) {
```
Comment (wldcordeiro) [inline src/utils/result-list.js]: I want to see if I can break this  out of `Autocomplete` and `SearchBar` so it's easily shared.

File: src/components/Editor/SearchBar.js [hunk 2]
```diff
@@ -345,7 +390,7 @@ const SearchBar = React.createClass({
     );
   },
 
-  renderFunctionSearchToggle() {
+  renderSearchTypeToggle() {
```
Comment (wldcordeiro) [inline src/components/Editor/SearchBar.js]: Prepping for when we have variable/class search types.

File: src/components/Editor/SearchBar.js [hunk 3]
```diff
@@ -266,41 +304,67 @@ const SearchBar = React.createClass({
...
+      }
+      e.preventDefault();
+    } else if (e.key === "Tab") {
+      this.closeSearch(e);
+      e.preventDefault();
+    }
+  },
+
+  // Renderers
+  buildSummaryMsg() {
```
Comment (wldcordeiro) [inline src/components/Editor/SearchBar.js]: Need some work here to have a summary for function search too.

File: src/components/Editor/SearchBar.js [hunk 4]
```diff
@@ -266,41 +304,67 @@ const SearchBar = React.createClass({
...
-    );
+    return L10N.getFormatStr("editor.searchResults", index + 1, count);
   },
 
-  renderSvg() {
-    const { searchResults: { count }} = this.props;
-
-    if (count == 0 && this.props.query.trim() != "") {
-      return Svg("sad-face");
+  buildPlaceHolder() {
```
Comment (wldcordeiro) [inline src/components/Editor/SearchBar.js]: Not sure it's necessary to change the placeholder but thought indicating the search type was good.

**Comment by wldcordeiro** [comment]
Things left to address:

1. Not doing normal code searching when function search is enabled
2. General clean-up of the code for simplicity/readability
3. Some styling bits for function search results
4?. Possibly axe `Autocomplete` in favor of reusing pattern from `SearchBar` in `SourceSearch`

**Comment by jasonLaster** [comment]
Hmm, this is cool. One thought is that we can minimize the risk a bit by keeping `Autocomplete` and making a new `Autocomplete2`. That way, we don't risk Cmd+P... but lets see where we're at soon. 

I'll pull it down and play

**Comment by jasonLaster** [comment]
Hmm, after playing with it seems like this refactor is pretty solid for source and editor search. I see some scrolling issues, but that's not a big deal

**Comment by jasonLaster** [comment]
This works well enough for the primary use cases that we can merge it and then focus on the function and modifier search.

**Pull request merged**